src/ray/core_worker/core_worker.cc

```diff
@@ -1106,23 +1106,71 @@ Status CoreWorker::CreateOwned(const std::shared_ptr<Buffer> &metadata,
                                const size_t data_size,
                                const std::vector<ObjectID> &contained_object_ids,
                                ObjectID *object_id, std::shared_ptr<Buffer> *data,
-                               bool created_by_worker) {
+                               bool created_by_worker,
+                               const std::shared_ptr<rpc::Address> owner_address) {
+  if (options_.is_local_mode) {
+    RAY_CHECK(owner_address == nullptr)
+        << "Specifying owner in `ray.put` is not allowed when running in local mode.";
+  }
   *object_id = ObjectID::FromIndex(worker_context_.GetCurrentTaskID(),
                                    worker_context_.GetNextPutIndex());
-  reference_counter_->AddOwnedObject(*object_id, contained_object_ids, rpc_address_,
-                                     CurrentCallSite(), data_size + metadata->Size(),
-                                     /*is_reconstructable=*/false,
-                                     NodeID::FromBinary(rpc_address_.raylet_id()));
-  if (options_.is_local_mode ||
-      (RayConfig::instance().put_small_object_in_memory_store() &&
-       static_cast<int64_t>(data_size) < max_direct_call_object_size_)) {
+  rpc::Address real_owner_address =
+      owner_address != nullptr ? *owner_address : rpc_address_;
+  bool owned_by_us = WorkerID::FromBinary(real_owner_address.worker_id()) ==
+                     WorkerID::FromBinary(rpc_address_.worker_id());
+  auto status = Status::OK();
+  if (owned_by_us) {
+    reference_counter_->AddOwnedObject(*object_id, contained_object_ids, rpc_address_,
+                                       CurrentCallSite(), data_size + metadata->Size(),
+                                       /*is_reconstructable=*/false,
+                                       NodeID::FromBinary(rpc_address_.raylet_id()));
+  } else {
+    // Because in the remote worker's `HandleAssignObjectOwner`,
+    // a `WaitForRefRemoved` RPC request will be sent back to
```

Comment:
It might have a race condition related to this [issue](https://github.com/ray-project/ray/issues/16688) since WaitForRefRemoved right now is an async call. After assigned object owner is returned, it doesn't guarantee the call is sent to this node.

Reply:
I think that's okay in this case. WaitForRefRemoved can arrive late.